### PR TITLE
feat(confidential): noir primitives library for confidential token circuits

### LINKED ISSUE #699: Noir primitives library for confidential token wrapper circuits
## Context

The Confidential Token Wrapper (see the internal design doc — link available in the tracking workspace) uses six Noir/UltraHonk circuits that share a common set of cryptographic primitives: Pedersen commitments over Grumpkin, Poseidon2 with typed domain separators, ECDH derivation, range checks for `[0, 2^128)`, and key-derivation helpers. Without a shared library, each circuit author will reimplement these — guaranteeing drift, audit pain, and a higher risk of subtle constraint differences between otherwise-identical operations.

This ticket establishes that library as the foundation all six circuits build on. It must land before circuit work begins.

## Scope

Build a single `circuits/lib/` Noir crate exposing every primitive the six circuits in §7 of the internal design doc reference. Concretely, in one ticket:

- **Generators**: `G`, `H` Grumpkin generators as `EmbeddedCurvePoint` globals (design doc §2.1, §10.4) with no known DL relation, documented provenance.
- **Pedersen commitment**: `commit(v: Field, r: Field) -> EmbeddedCurvePoint` via `multi_scalar_mul([G, H], [v, r])` (§2.3, §10.4).
- **Poseidon2 wrapper with typed domain separators**: one canonical hash function per domain, covering all 11 constants in §13 of the design doc (`δ_vk`, `δ_dvk`, `δ_spend_r`, `δ_tx_blind`, `δ_tx_amount`, `δ_enc_bal`, `δ_enc_allow`, `δ_allow_r`, `δ_esc_dvk`, `δ_aud_tx`, `δ_aud_bal`).
- **ECDH derivation**: `ecdh(scalar, point) -> shared_secret_x` matching §2.4.
- **Range check**: `assert_range_128(x)` enforcing `x ∈ [0, 2^128)` via the native `u128` cast pattern in §2.6, reused by W4 / T4 / O4 / S4 / V9.
- **Key derivation helpers**:
  - `vk_from_sk(sk, wrap) = Poseidon2(δ_vk, sk, wrap)` (§4.2)
  - `pvk_from_vk(vk) = vk · H` (§4.3)
  - `dvk_from_vk_op(vk, op_i) = Poseidon2(δ_dvk, vk, op_i)` (§4.4)
- **Encrypted scalar helpers**: one fn per ciphertext kind used in §7 (`encrypt_amount`, `encrypt_balance`, `encrypt_allowance`, `encrypt_auditor_tx`, `encrypt_auditor_bal`, `derive_spend_r`, `derive_allow_r`, `derive_tx_blind`).
- **Test vectors**: JSON fixtures under `circuits/lib/testdata/` that pin the exact output of each primitive for a fixed input. These fixtures are also the cross-language contract for the SDK — the SDK must reproduce them byte-for-byte.

Out of scope: the 6 circuits themselves; on-chain point arithmetic (separate ticket).

## Key Questions
- Where do `G` and `H` come from? (Hash-to-curve from a public seed vs. Aztec-provided generators — needs decision before merge.)
- Do we need a `verify_pedersen_opening(C, v, r)` helper, or do circuits inline that as `assert C == commit(v, r)`?
- Should the test-vector fixtures live in the circuits repo or in a shared `test-vectors/` package consumed by both Noir and the SDK?

## Proposed Approach
Build the library in a single Noir crate so all symbols compile together and the test-vector harness exercises every helper from one entry point.

1. Scaffold `circuits/lib/` as a Noir crate; declare generators with documented provenance (resolve the open question above before merging).
2. Implement Pedersen `commit`, ECDH, and `assert_range_128`. Add unit tests that build, e.g., `commit(0, 0) == O` and `commit(v, r) + commit(v', r') == commit(v+v', r+r')` to confirm homomorphism.
3. Define a single `enum Domain { Vk, Dvk, SpendR, ... }` (or an explicit constant module) and a single `poseidon_with_domain(d, inputs)` wrapper. Every higher-level helper goes through it — no caller hashes raw.
4. Build the key-derivation and encrypted-scalar helpers on top of the Poseidon wrapper. Each function maps 1:1 to a constraint line in §7 of the design doc — link them in doc-comments.
5. Generate canonical test vectors with a deterministic input set (seeded RNG) and serialize to JSON. Add a `cargo test` (or `nargo test`) harness that verifies vector reproduction. Publish the vectors as build artifacts so the SDK and on-chain code can pull them.
6. Run a constraint-count benchmark per primitive (`nargo info`) and post the numbers in the PR description — this baselines circuit-cost analysis for §10.3 of the design doc.

## Known References
- Internal design doc — §2 (Preliminaries), §4 (Key Hierarchy), §10.4 (Noir Primitives), §13 (Domain Separation Constants). Link available in the tracking workspace.
- Noir stdlib `std::embedded_curve_ops`
- [Poseidon2 paper](https://eprint.iacr.org/2023/323) and the Barretenberg / Noir stdlib references listed in §2.5 of the design doc

## Acceptance Criteria
- [ ] Implementation matches architecture brief
- [ ] Tests cover main flows and edge cases
- [ ] Review/audit findings addressed (if applicable)
- [ ] PR links back to this issue

### PR DESCRIPTION
**Pull request by brozorec**
## Summary

Establishes the Noir primitives library that all six confidential-token-wrapper circuits (Register, Withdraw, Transfer, OperatorTransfer, SetOperator, RevokeOperator -- §7 of the design doc) will depend on. Lands the lib, fixtures, gadget circuits, CI, and documentation.

Closes #699.

### `packages/tokens/src/confidential/circuits/lib/`

The canonical primitive crate. All hashing funnels through one `poseidon_with_domain` entry point so the domain tag is always the *first* absorbed element -- no caller hashes raw.

| Primitive | Reference |
|:--|:--|
| `G`, `H` generators (no known DL relation) | Barretenberg `derive_generators("DEFAULT_DOMAIN_SEPARATOR", 0..2)`; provenance asserted at runtime by `print_generators` test |
| `commit(v, r)` | Section 2.3 |
| `scalar_mul(s, P)`, `ecdh(s, P) -> Field` | Section 2.4 |
| `assert_range_128(x)` | Section 2.6 |
| `poseidon_with_domain(d, inputs)` | The funnel; 11 domain tags from Section 13 |
| `vk_from_sk(sk, wrap)` | Section 4.2 -- **3-input, `wrap`-bound** |
| `pvk_from_vk(vk)` | Section 4.3 |
| `dvk_from_vk_op(vk, op_i)` | Section 4.4 |
| `derive_spend_r`, `derive_allow_r`, `derive_tx_blind` | W5/T10/S9/V6, S6/O3/O10, T7/O7 |
| `encrypt_amount`, `encrypt_balance`, `encrypt_allowance`, `encrypt_auditor_tx`, `encrypt_auditor_bal`, `encrypt_esc_dvk` | T9/O9, W7/T12/S11/V8, S8/O12, T_a2..S_a3, W_a3..O_a5, S12 |

### `lib/testdata/*.json`

16 JSON fixtures pinning every primitive's output for a fixed input set. These are the cross-language contract: any SDK consumer (TS, Rust, on-chain integration tests) is correct *iff* it reproduces every output byte-for-byte. The lib's `fixtures_match_testdata` test guards parity from the Noir side.

### `gadgets/<primitive>/`

One minimal `bin` crate per primitive (`commit`, `ecdh`, `vk_from_sk`, `encrypt_amount`, `range_128`, `poseidon_with_domain`). They exist because `nargo info` measures `main`, not `#[test]`, so per-primitive constraint counts require per-primitive mains. Each gadget pins a fixture-matching `#[test]` so the bench code can't silently drift from the lib.

### `.github/workflows/noir.yml` + `circuits/constraints.baseline`

CI runs `nargo check && nargo test`, then diffs normalized `nargo info` output against the checked-in baseline. Drift fails the build with a step-summary panel showing exactly which command to run to regenerate. First Noir CI job in the repo; uses `noirup` to pin nargo 1.0.0-beta.11.

### `Architecture.md`

New "Noir Circuits" section documenting the layout convention (`circuits/{lib, gadgets, <operation>}/`), why the Noir tree is its own workspace independent of Cargo, and the commands cheat sheet.

## Constraint baseline (`nargo info`)

| Gadget | ACIR Opcodes | Brillig Opcodes |
|:--|--:|--:|
| `commit` | 23 | 63 |
| `ecdh` | 12 | 63 |
| `range_128` | 14 | 17 |
| `vk_from_sk` (Poseidon, 3 inputs) | 4 | 0 |
| `encrypt_amount` (Poseidon + add) | 4 | 0 |
| `poseidon_with_domain` (2 inputs) | 3 | 0 |

Full machine-readable table lives in `constraints.baseline` and is enforced by the CI diff.

## Decisions worth flagging for review

- **G/H provenance.** Aztec/Barretenberg standard Pedersen generators via `derive_generators`. The previous prototype used `H = 7·G` -- known DL relation, breaks binding. Hardcoded coords here are re-derived at runtime by `print_generators` so drift from upstream is caught immediately.
- **`vk = Poseidon2(δ_vk, sk, wrap)` includes `wrap`.** The previous prototype dropped it; without `wrap` the proofs aren't wrapper-context-bound (constraints R2/W2/T2/S2/V2). The `vk_wrap_binding` test guards this.
- **Open question Q2 (`verify_pedersen_opening` helper).** Skipped. Existing circuits inline `assert C == commit(v, r)` as two `.x`/`.y` checks (W3, T3); a helper would save four lines per call site at the cost of a function whose contract diverges subtly from `commit`. Easy to add later if usage patterns warrant.
- **Single `poseidon_with_domain` funnel.** Every higher-level helper routes through it; arity-specific sponge wrappers are module-private. Audit surface is one function.

## Out of scope (per the issue)

- The 6 circuits themselves -- follow-up tickets.
- On-chain point arithmetic -- separate ticket.
- Reconciling the draft circuits in the `confidential-test` repo to consume this lib -- follow-up.

## Test plan

- [x] `nargo test` -- 23 tests pass locally (15 in lib, 4 in gadget fixture-matchers, 3 in range_128 boundary, 1 in poseidon_with_domain).
- [x] `nargo info` -- per-primitive constraint counts captured in `constraints.baseline`.
- [x] `nargo check` -- workspace type-checks cleanly.
- [x] `cargo check --package stellar-tokens` -- Cargo workspace unaffected by the new Noir tree.
- [x] Noir CI workflow runs green on this PR (first time exercising it on a GitHub runner).
- [x] Reviewer verifies `print_generators` output matches the hardcoded G/H -- one command: `nargo test print_generators --package stellar_confidential_lib --show-output`.

## Summary by CodeRabbit

* **New Features**
  * Added Noir circuits for confidential token operations with core cryptographic primitives (commitment, ECDH, encryption, range proofs, key derivation).
  * Added circuit gadgets and shared cryptographic library for confidential token wrapper functionality.

* **Documentation**
  * Added architecture documentation for Noir circuits organization and commands.
  * Added test fixture documentation and cross-language test vectors for validation.

* **Tests**
  * Added GitHub Actions workflow for circuit validation and constraint tracking with baseline comparison.

[Review Change Stack](https://app.coderabbit.ai/change-stack/OpenZeppelin/stellar-contracts/pull/712?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .github/workflows/noir.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        run: |
+          curl -L https://raw.githubusercontent.com/noir-lang/noirup/main/install | bash
+          echo "$HOME/.nargo/bin" >> "$GITHUB_PATH"
+        working-directory: ${{ github.workspace }}
+
+      - name: Pin nargo version
+        run: noirup -v "$NARGO_VERSION"
+
+      - name: nargo --version
+        run: nargo --version
```
Comment (ozgunozerk) [inline .github/workflows/noir.yml]: what are the uses of having these steps in a workflow?

File: packages/tokens/src/confidential/circuits/lib/testdata/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+## Regenerating
+
+The fixtures are produced from the Noir lib itself:
+
+```bash
+nargo test print_fixtures --package stellar_confidential_lib --show-output
+```
+
+Capture the printed values into the corresponding `*.json` file (one per primitive). The `fixtures_match_testdata` test in `lib.nr` asserts the lib still produces these exact values; if it fails, either the fixtures were regenerated without updating the test (out-of-sync) or a primitive's semantics changed (breaking the cross-language contract — bump a version).
```
Comment (ozgunozerk) [inline packages/tokens/src/confidential/circuits/lib/testdata/README.md]: `Capture the printed values into the corresponding `*.json` file (one per primitive).` 

This is my first time reading this, and it is the readme file. I don't even know the context of which file should I run, and which values should I print. I think more context here may be beneficial

File: Architecture.md [hunk 3]
```diff
@@ -277,6 +277,37 @@ The non-fungible token implementation is designed to be compatible with existing
...
+Noir artifacts for a token kind live under `packages/tokens/src/<kind>/circuits/`:
+
+```text
+packages/tokens/src/<kind>/circuits/
+├── Nargo.toml               # Nargo workspace manifest
+├── lib/                     # Shared primitives (type = "lib", no `main`)
+│   ├── src/lib.nr
+│   └── testdata/*.json      # cross-language test vectors
+├── gadgets/<primitive>/     # one tiny `bin` crate per primitive
+│   └── src/main.nr
```
Comment (ozgunozerk) [inline Architecture.md]: I was reviewing this PR, and found on my own (i'm still not sure about it), that `main.nr` acts as a test in this context, am I right? 

Let's also put an exaplanation here on what this `src/main.rs` means, and what purpose does it serve

File: Architecture.md [hunk 4]
```diff
@@ -277,6 +277,37 @@ The non-fungible token implementation is designed to be compatible with existing
...
+└── <operation>/             # full circuits (e.g. register, transfer)
+    └── src/main.nr
```
Comment (ozgunozerk) [inline Architecture.md]: same comment here

File: Architecture.md [hunk 5]
```diff
@@ -277,6 +277,37 @@ The non-fungible token implementation is designed to be compatible with existing
...
+
+```text
+packages/tokens/src/<kind>/circuits/
+├── Nargo.toml               # Nargo workspace manifest
+├── lib/                     # Shared primitives (type = "lib", no `main`)
+│   ├── src/lib.nr
+│   └── testdata/*.json      # cross-language test vectors
+├── gadgets/<primitive>/     # one tiny `bin` crate per primitive
+│   └── src/main.nr
+└── <operation>/             # full circuits (e.g. register, transfer)
```
Comment (ozgunozerk) [inline Architecture.md]: I didn't notice any `operation` subfolder in this PR. Will they be introduced later? If so, ok. If not and if this is an artifact, shall we remove this entry?

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+    v_new + poseidon_with_domain(domain::AUD_BAL, [s_a, sigma])
+}
+
+/// Delegation-key escrow mask for the operator handoff at `set_operator`. Section 7.11.
+/// `escrowed_dvk = dvk + Poseidon2(ESC_DVK, s, op_i)` where `s = (r_e * Y_op).x`.
+pub fn encrypt_esc_dvk(dvk: Field, s: Field, op_i: Field) -> Field {
+    dvk + poseidon_with_domain(domain::ESC_DVK, [s, op_i])
+}
+
+// ################## TESTS ##################
```
Comment (ozgunozerk) [inline packages/tokens/src/confidential/circuits/lib/src/lib.nr]: I think putting the tests in somewhere else is a great idea. This file is already too long and a bit scary <3 

On top of that, I have another question regarding tests. These are unit tests, right? But I believe the `main.nr` ones are also acting as unit tests. Probably, there is another detail in this context, and mapping them to the concept `unit test` is not covering every thing. Can you explain to me why do we have both tests in here and `main.nr` for each primitive?

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr [hunk 7]
```diff
@@ -0,0 +1,704 @@
+//! # Confidential Token Wrapper -- Noir Primitives
+//!
+//! Shared cryptographic primitives consumed by the six circuits described in
+//! Section 7 of the design document (Register, Withdraw, Transfer, OperatorTransfer,
+//! SetOperator, RevokeOperator).
```
Comment (ozgunozerk) [inline packages/tokens/src/confidential/circuits/lib/src/lib.nr:5]: Throughout this PR, I think mentioning/referring to the design document, and its `Section`s , is not a good practice when we don't have the design document in the repo. I suggest reworking all the non-code part of this PR w.r.t this

Comment (ozgunozerk) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr:5]: Or, we can put the design document here, I'm open for that discussion

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+//!
+//! Every constraint that involves a Pedersen commitment, a Poseidon2 hash with
+//! domain separation, an ECDH derivation, or a 128-bit range check funnels
+//! through one of the helpers exported here. Circuits never hash raw --
+//! `poseidon_with_domain` is the only Poseidon entry point.
+//!
+//! ## Curve
+//!
+//! Grumpkin `y^2 = x^3 - 17` over `Fr_{BN254}`. Both generators below come
```
Comment (ozgunozerk) [inline packages/tokens/src/confidential/circuits/lib/src/lib.nr]: I tried googling this value `Fr_BN254`, the results were not that helpful. Can we give a good resource/link/explanation for intro on this, so that the inline docs become more helpful?

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+/// hash-to-curve on the domain separator + index).
+pub global H: EmbeddedCurvePoint = EmbeddedCurvePoint {
+    x: 0x054aa86a73cb8a34525e5bbed6e43ba1198e860f5f3950268f71df4591bde402,
+    y: 0x209dcfbf2cfb57f9f6046f44d71ac6faf87254afc7407c04eb621a6287cac126,
+    is_infinite: false,
+};
+
+// ################## CONSTANTS ##################
+
+/// Poseidon2 sponge IV multiplier: `iv = (input_length as Field) * 2^64`.
```
Comment (ozgunozerk) [inline packages/tokens/src/confidential/circuits/lib/src/lib.nr]: i believe IV stands for `Initialization Vector`, right? Can we use the full name at least for the first time of the abbreviations?

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr [hunk 10]
```diff
@@ -0,0 +1,704 @@
...
+/// Matches Noir stdlib's `Poseidon2Hasher` and Barretenberg's Poseidon2 sponge.
+global POSEIDON2_IV_BASE: Field = 18446744073709551616; // 2^64
+
+/// Domain separation tags (Section 13 of the design document). Module-private --
+/// callers reach them through the typed `poseidon_with_domain` helpers below.
+mod domain {
```
Comment (ozgunozerk) [inline packages/tokens/src/confidential/circuits/lib/src/lib.nr:91]: i'm not sure about the names for the constants. For example, can't we use VIEW_KEY instead of VK? Would be less confusing. 
Also, SPEND_R, why not SPEND_BALANCE? what does R corresponds to? Do we expect from ourselves and other devs to memorize these shorthands?

**Review by ozgunozerk — CHANGES_REQUESTED** [review]

File: Architecture.md (hunk 5 above)
Comment (brozorec) [reply Architecture.md]: yes, they are to be introduced with every new circuit

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr (hunk 7 above)
Comment (brozorec) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr:5]: will get that added to the repo 👍

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr (hunk 6 above)
Comment (brozorec) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr]: good point, tests in each primitive are redundant; will put tests in a another mod

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr (hunk 10 above)
Comment (brozorec) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr:91]: They follow the notation from the design doc, but I agree with you that we better use full names.

File: packages/tokens/src/confidential/circuits/lib/src/lib.nr (hunk 6 above)
Comment (ozgunozerk) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr]: Sounds good! This question is still unanswered for me though:

> These are unit tests, right? But I believe the main.nr ones are also acting as unit tests. Probably, there is another detail in this context, and mapping them to the concept unit test is not covering every thing. Can you explain to me why do we have both tests in here and main.nr for each primitive?

Comment (ozgunozerk) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr]: I see that the tests in the `main.nr` files are removed, but what is the use for `main.nr` files then?

Comment (ozgunozerk) [reply packages/tokens/src/confidential/circuits/lib/src/lib.nr]: Ok, I think this is the explanation put in another commit:

> - **Gadget `main`s** are measurement-only. Each one wraps a single lib primitive in the smallest circuit that exercises it, so `nargo info` can report per-primitive constraint costs.
> - **Operation `main`s** *are* the production circuits — one per proof type listed in the design doc's §7. They consume on-chain state as public inputs and enforce the constraint set that the wrapper contract relies on.

Resolving this one 👍

**Review by ozgunozerk — APPROVED** [review]
